qutip/core/data/add.pyx

```diff
@@ -322,11 +426,29 @@ add.__doc__ =\
     scalar.
     """
 add.add_specialisations([
-    (Dense, Dense, Dense, add_dense),
+    (COO, COO, COO, add_coo),
     (CSR, CSR, CSR, add_csr),
+    (Dense, Dense, Dense, add_dense),
     (Dia, Dia, Dia, add_dia),
 ], _defer=True)
 
+from .convert import to as _to
+
+cpdef Dense iadd_dense_data(Dense left, Data right, double complex scale=1):
+    cdef Data out = add(left, right)
+    cdef idxint size = left.shape[0]* left.shape[1]
+    cdef Dense out_dense = _to(Dense, out)
+    memcpy(left.data, out_dense.data, size*sizeof(double complex))
+    return left
+
+cpdef Dense iadd_dense(Dense left, Data right, double complex scale=1):
```

Comment:
For function with multiple type: `{operation}_{first_dtype}_{second_dtype}_..._{out_dtype}`

```suggestion
cpdef Dense iadd_dense_data_dense(Dense left, Data right, double complex scale=1):
```